src/extensions/agent_modules/cloud_communication/adps2_mqtt_client/src/adps2_mqtt_client_module.c

```diff
@@ -132,76 +113,182 @@ enum ADPS_ERROR
+    const char* status = json_object_get_string(json_object(root_value), "status");
+    if (status == NULL)
     {
-        Log_Info("on_connect: %s", mosquitto_connack_string(reason_code));
+        Log_Error("Failed to get status from JSON payload");
+        goto done;
     }
 
-    //   if (reason_code != 0)
-    //   {
-    //     keep_running = 0;
-    //     /* If the connection fails for any reason, we don't want to keep on
-    //      * retrying in this example, so disconnect. Without this, the client
-    //      * will attempt to reconnect. */
-    //     int rc;
-    //     if ((rc = mosquitto_disconnect_v5(mosq, reason_code, NULL)) != MOSQ_ERR_SUCCESS)
-    //     {
-    //       Log_Error("Failure on disconnect: %s", mosquitto_strerror(rc));
-    //     }
-    //   }
-}
+    if (strncmp(status, "assigned", 8) == 0)
+    {
+        Log_Info("Device is registered.");
+        SetRegisterState(ADPS_REGISTER_STATE_REGISTERED, "received assigned status");
+    }
+    else if (strncmp(status, "assigning", 9) == 0)
+    {
+        Log_Info("Device is registering.");
+        SetRegisterState(ADPS_REGISTER_STATE_REGISTERING, "received assigning status");
+    }
+    else
+    {
+        Log_Error("Unknown status: %s", status);
+        SetRegisterState(ADPS_REGISTER_STATE_FAILED, "received unknown status");
+        // TODO (nox-msft) : determine whether to retry or bail
+    }
 
-/* Callback called when the broker has received the DISCONNECT command and has disconnected the
- * client. */
-void ADPS_OnDisconnect(struct mosquitto* mosq, void* obj, int rc, const mosquitto_property* props)
-{
-    connection_ok = 0;
-    Log_Info("on_disconnect: reason=%s", mosquitto_strerror(rc));
+done:
+    return result;
 }
 
 /* Callback called when the client receives a message. */
-void ADPS_OnMessage(
+void ADPS_MQTT_Client_Module_OnMessage(
     struct mosquitto* mosq, void* obj, const struct mosquitto_message* msg, const mosquitto_property* props)
 {
+    bool result = false;
+    char** topics = NULL;
+    int count = 0;
     Log_Info("on_message: Topic: %s; QOS: %d; mid: %d", msg->topic, msg->qos, msg->mid);
 
-    //ADPS_MQTT_CLIENT_MODULE_STATE* state = (ADPS_MQTT_CLIENT_MODULE_STATE*)obj;
+    if (msg != NULL && msg->topic != NULL)
+    {
+        time_t nowTime = ADUC_GetTimeSinceEpochInSeconds();
+
+        int mqtt_ret = mosquitto_topic_matches_sub(ADPS_MQTT_TOPIC_REGISTRATIONS_RESULT, msg->topic, &result);
+        if (mqtt_ret != MQTT_RC_SUCCESS || !result)
+        {
+            goto done;
+        }
+
+        // Get http status code from topic sub segment
+        // e.g., "$dps/registrations/res/200/?$rid=1&..."
+        mqtt_ret = mosquitto_sub_topic_tokenise(msg->topic, &topics, &count);
+        if (mqtt_ret != MQTT_RC_SUCCESS && count <= 3)
+        {
+            // Cant get the http status code.
+            goto done;
+        }
 
-    // TODO (nox-msft) : Handle the message.
+        int http_status_code = atoi(topics[3]);
```

Comment:
I may have missed it, but Where is the validation code to ensure path segment 0 is "$dps", path segment 1 is "registrations", etc?